### PR TITLE
sync discovery service in controller

### PR DESCRIPTION
**Pull request by aylei**
Signed-off-by: Aylei <rayingecho@gmail.com>

close #1322 

### Check List 

Tests 

 - Unit test
 - E2E test

Code changes

 - Has Helm charts change
 - Has Go code change

### Does this PR introduce a user-facing change?:
 
 ```release-note
`tidb-controller-manager` will automatically create and reconcile the tidb-discovery-service now
 ```

@weekface @cofyc @tennix @DanielZhangQD PTAL

### CODE REVIEW

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 1]
```diff
@@ -74,6 +78,8 @@ var _ = ginkgo.Describe("[tidb-operator] TiDBCluster", func() {
 		framework.ExpectNoError(err, "failed to create clientset")
 		asCli, err = asclientset.NewForConfig(config)
 		framework.ExpectNoError(err, "failed to create clientset")
+		genericCli, err = client.New(config, client.Options{Scheme: scheme.Scheme})
```
Comment (cofyc) [inline tests/e2e/tidbcluster/tidbcluster.go]: no need to registry k8s scheme?

Comment (aylei) [reply tests/e2e/tidbcluster/tidbcluster.go]: The new `scheme.go` has `k8s.io`, `pingcap.com` and `asts` registered.

File: pkg/controller/controller_utils.go [hunk 2]
```diff
@@ -337,3 +348,38 @@ func (rt *RequestTracker) SetRequests(requests int) *RequestTracker {
...
+// InferObjectKind infers the object kind
+func InferObjectKind(obj runtime.Object) (schema.GroupVersionKind, error) {
+	gvks, _, err := scheme.Scheme.ObjectKinds(obj)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+	if len(gvks) != 1 {
+		return schema.GroupVersionKind{}, fmt.Errorf("Object %v has ambigious GVK", obj)
+	}
+	return gvks[0], nil
```
[the remark this answers is not in the document]
Comment (aylei) [reply pkg/controller/controller_utils.go]: then `len(gvks)` is 0

File: pkg/controller/equality.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	podSpec := &corev1.PodSpec{}
+	err := json.Unmarshal([]byte(applied), podSpec)
+	if err != nil {
+		return nil, err
+	}
+	return podSpec, nil
+}
+
+// DeploymentPodSpecChanged checks whether the new deployment differs with the old one's last-applied-config
+func DeploymentPodSpecChanged(new *appsv1.Deployment, old *appsv1.Deployment) bool {
```
[the remark this answers is not in the document]
Comment (aylei) [reply pkg/controller/equality.go]: Good catch!

File: pkg/manager/member/tidb_discovery_manager.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		return controller.RequeueErrorf("error creating or updating discovery serviceaccount: %v", err)
+	}
+	_, err = m.ctrl.CreateOrUpdateRoleBinding(tc, &rbacv1.RoleBinding{
+		ObjectMeta: meta,
+		Subjects: []rbacv1.Subject{{
+			Kind: "ServiceAccount",
+			Name: meta.Name,
+		}},
+		RoleRef: rbacv1.RoleRef{
+			Kind:     "Role",
```
[the remark this answers is not in the document]
Comment (aylei) [reply pkg/manager/member/tidb_discovery_manager.go]: Seems like there's no Role const defined in kubernetes, and kubernetes /kubernetes uses literal too https://github.com/kubernetes/kubernetes/blob/5135a13e77d9863fb73292d7b793fbdf23991bbe/pkg/apis/rbac/validation/validation_test.go#L67

I suggest to keep it same as the k8s code

**Comment by cofyc** [comment]
/run-e2e-test
the kind failed to create the k8s cluster for E2E v1.16.3, I'll report this issue to kind

**Pull request merged**